test/tests/functional/pbs_soft_walltime.py

```diff
@@ -687,3 +709,336 @@ def test_preempt_order(self):
+
+    def test_soft_less_cput(self):
+        """
+        Test that soft_walltime has no impact on cput enforcement limit
+        """
+
+        script = []
+        script += ['i=0;\n']
+        script += ['while [ $i -ne 0 ] || sleep 0.125;\n']
+        script += ['do i=$(((i+1) % 10000 ));\n']
+        script += ['done\n']
+        j1 = Job(TEST_USER, {'Resource_List.cput': 5})
+        j1.create_script(body=script)
+        jid = self.server.submit(j1)
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 300})
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        # verify that job is deleted when cput limit is reached
+        time.sleep(10)
+        self.server.expect(JOB, 'queue', op=UNSET, id=jid)
+        self.server.log_match(
+            "%s.* resources_used.cput=00:00:05" % jid, regexp=True)
+
+    def test_soft_walltime_resv(self):
+        """
+        Submit a job with soft walltime inside a reservation
+        """
+
+        now = int(time.time())
+
+        a = {'Resource_List.ncpus': 1, 'reserve_start': now + 5,
+             'reserve_end': now + 10}
+        R = Reservation(TEST_USER, attrs=a)
+        rid = self.server.submit(R)
+        self.scheduler.log_match(
+            rid + ';Reservation Confirmed', max_attempts=5)
+        r1 = rid.split('.')[0]
+
+        j1 = Job(TEST_USER, attrs={ATTR_queue: r1})
+        jid = self.server.submit(j1)
+        # Set soft walltime to greater than reservation end time
+        self.server.alterjob(jid, {'Resource_List.soft_walltime': 300})
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 300}, id=jid)
+
+        # verify that the job gets deleted when reservation ends
+        self.server.expect(
+            JOB, 'queue', op=UNSET, id=jid, offset=10, max_attempts=10)
+
+    def test_restart_server(self):
+        """
+        Test that on server restart soft walltime value is reset
+        """
+
+        hook_body = \
+            """import pbs
+e = pbs.event()
+e.job.Resource_List["soft_walltime"] = 5
+e.accept()
+"""
+        a = {'event': 'queuejob', 'enabled': 'True'}
+        self.server.create_import_hook("que", a, hook_body)
+
+        J = Job(TEST_USER)
+        jid = self.server.submit(J)
+
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 5}, id=jid)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        time.sleep(6)
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+
+        self.server.expect(JOB, {'estimated.soft_walltime':
+                                 (MATCH_RE, '00:00:10|10')}, id=jid)
+
+        self.server.restart()
+
+        self.server.expect(JOB, {'Resource_List.soft_walltime': 5}, id=jid)
+
+    def test_resv_job_soft_hard2(self):
+        """
+        Test that a job with soft and hard walltime will not conflict with
+        reservtion if hard walltime is less that reservation start time.
```

Comment:
you might be right the code wise but this is a blackbox test and in black box world it is an important integration check.

Reply:
Just a thought, in the test case 'test_resv_job_soft_hard', we check that a job with hard walltime > reservation start time is not run even though its soft walltime is < reservation start time. Why not just add a job alter in the same test to reduce the hard walltime to be < reservation start time as well, which will make the job run. I think that will test the same thing that you are testing here. What do you think?

Reply:
Thanks @agrawalravi90. you have a valid suggestion but in general it is recommended to create different test cases for different objective otherwise if one is broken whole test will block. For this case I have no preference and can go otherwise if you have a strong feeling that we should do that.